File: FREQNESS_Toolbox/python/src/freqness/FREQNESS_ExponentialDK.py
"""Model exponential decay of FREQ-NESS eigenvalues across frequency."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping
import warnings

import numpy as np
from numpy.typing import ArrayLike, NDArray


FloatArray = NDArray[np.float64]


@dataclass(slots=True)
class FREQNESSExponentialGoodFit:
    """Goodness-of-fit information returned by FREQNESS_ExponentialDK."""

    R2: FloatArray


def _field(container: Any, name: str, *, required: bool = True) -> Any:
    if isinstance(container, Mapping):
        if name in container:
            return container[name]
    elif hasattr(container, name):
        return getattr(container, name)
    if required:
        raise ValueError(f"FREQ.{name} is required")
    return None


def _eigenspectrum_array(FREQ: Any) -> FloatArray:
    values = np.asarray(_field(FREQ, "evals"))
    if values.size == 0:
        raise ValueError("FREQ.evals is empty")
    if not np.issubdtype(values.dtype, np.number) or np.iscomplexobj(values):
        raise TypeError("FREQ.evals must contain real numeric values")
    values = np.asarray(values, dtype=float)
    if values.ndim == 1:
        values = values[:, np.newaxis, np.newaxis]
    elif values.ndim == 2:
        values = values[:, :, np.newaxis]
    elif values.ndim != 3:
        raise ValueError(
            "FREQ.evals must be 1D, 2D, or 3D "
            "(components, frequencies[, participants])"
        )
    if min(values.shape) == 0:
        raise ValueError("FREQ.evals is empty")
    return values


def _frequency_axis(FREQ: Any, nfrequencies: int) -> tuple[FloatArray, str, bool]:
    raw = _field(FREQ, "frex", required=False)
    if raw is not None:
        frequencies = np.asarray(raw, dtype=float)
        if (
            frequencies.ndim in (1, 2)
            and (frequencies.ndim == 1 or 1 in frequencies.shape)
            and frequencies.size == nfrequencies
            and np.all(np.isfinite(frequencies))
        ):
            return frequencies.reshape(-1).copy(), "Frequency (Hz)", True
    return (
        np.arange(1, nfrequencies + 1, dtype=float),
        "Frequency index",
        False,
    )


def _component_index(which_comp: Any, ncomponents: int) -> int:
    if which_comp is None:
        return 0
    valid = (
        isinstance(which_comp, (int, np.integer))
        and not isinstance(which_comp, (bool, np.bool_))
        and 1 <= int(which_comp) <= ncomponents
    )
    if not valid:
        warnings.warn(
            f"which_comp must be an integer between 1 and {ncomponents}. "
            "Defaulting to component 1.",
            UserWarning,
            stacklevel=3,
        )
        return 0
    return int(which_comp) - 1


def _fit_indices(
    x_axis: FloatArray,
    has_frequencies: bool,
    range2fit: ArrayLike | None,
) -> NDArray[np.int64]:
    if range2fit is None:
        return np.arange(x_axis.size, dtype=np.int64)
    requested = np.asarray(range2fit, dtype=float)
    if requested.size == 0:
        return np.arange(x_axis.size, dtype=np.int64)
    if not has_frequencies:
        raise ValueError("range2fit was provided in Hz, but FREQ.frex is missing")
    if requested.ndim != 1 or requested.size != 2:
        raise ValueError("range2fit must be a two-element vector such as [8, 12]")
    if not np.all(np.isfinite(requested)):
        raise ValueError("range2fit must contain finite values")
    if requested[0] > requested[1]:
        raise ValueError("range2fit boundaries must be in ascending order")

    first = int(np.argmin(np.abs(x_axis - requested[0])))
    last = int(np.argmin(np.abs(x_axis - requested[1])))
    if first > last:
        raise ValueError("range2fit does not map to an ascending frequency range")
    matched = np.array([x_axis[first], x_axis[last]])
    if not np.all(np.isclose(requested, matched, atol=1e-3, rtol=0.0)):
        warnings.warn(
            "Some requested frequencies in range2fit are not present in "
            "FREQ.frex. Using closest matches instead.",
            UserWarning,
            stacklevel=3,
        )
    return np.arange(first, last + 1, dtype=np.int64)


def _fit_exponential(
    x_values: FloatArray,
    y_values: FloatArray,
) -> tuple[float, float, float]:
    mask = (y_values > 0) & np.isfinite(y_values) & np.isfinite(x_values)
    if np.count_nonzero(mask) < 3:
        return np.nan, np.nan, np.nan

    x_fit = x_values[mask]
    y_fit = y_values[mask]
    design = np.column_stack([x_fit, np.ones(x_fit.size)])
    slope, intercept = np.linalg.lstsq(design, np.log(y_fit), rcond=None)[0]
    decay = float(-slope)
    amplitude = float(np.exp(intercept))
    predicted = amplitude * np.exp(-decay * x_fit)
    sse = float(np.sum((y_fit - predicted) ** 2))
    sst = float(np.sum((y_fit - np.mean(y_fit)) ** 2))
    r_squared = 1.0 - sse / sst if sst > 0 else np.nan
    return decay, amplitude, r_squared


def _nanmean(values: FloatArray, axis: int) -> FloatArray:
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", category=RuntimeWarning)
        return np.nanmean(values, axis=axis)


def _set_x_limits(axis: Any, x_axis: FloatArray) -> None:
    if x_axis.size == 1:
        padding = max(abs(float(x_axis[0])) * 0.05, 0.5)
        axis.set_xlim(x_axis[0] - padding, x_axis[0] + padding)
    else:
        axis.set_xlim(float(np.min(x_axis)), float(np.max(x_axis)))


def _style_axis(axis: Any, x_axis: FloatArray, xlabel: str, title: str) -> None:
    axis.set_xlabel(xlabel, fontweight="bold")
    axis.set_ylabel("Eigenvalue", fontweight="bold")
    axis.set_title(title, fontweight="bold")
    axis.grid(True, which="both", alpha=0.25)
    axis.spines["top"].set_visible(False)
    axis.spines["right"].set_visible(False)
    _set_x_limits(axis, x_axis)


def _plot_decay(
    eigenspectrum: FloatArray,
    x_axis: FloatArray,
    xlabel: str,
    fit_indices: NDArray[np.int64],
    component_index: int,
    decay_coefficients: FloatArray,
    amplitudes: FloatArray,
    group_decay: float,
    group_amplitude: float,
    *,
    plot_all: bool,
) -> list[Any]:
    try:
        import matplotlib.pyplot as plt
    except ImportError as error:
        raise ImportError(
            "FREQNESS_ExponentialDK plotting requires matplotlib. "
            "Install freqness[plot]."
        ) from error

    nsubjects = eigenspectrum.shape[1]
    colormap = plt.get_cmap("viridis")
    spectrum_color = colormap(0.12)
    fit_color = colormap(0.78)
    figures: list[Any] = []

    average = _nanmean(eigenspectrum, axis=1)
    if nsubjects > 1:
        sem = np.nanstd(eigenspectrum, axis=1, ddof=1) / np.sqrt(nsubjects)
    else:
        sem = None

    figure, axis = plt.subplots(figsize=(8, 5), constrained_layout=True)
    if sem is not None and not np.all(np.isnan(sem)):
        axis.fill_between(
            x_axis,
            average - sem,
            average + sem,
            color="black",
            alpha=0.1,
            linewidth=0,
        )
    spectrum_line = axis.plot(
        x_axis,
        average,
        color=spectrum_color,
        linewidth=2,
        label="Mean eigenspectrum",
    )[0]
    handles = [spectrum_line]
    if np.isfinite(group_decay) and np.isfinite(group_amplitude):
        fit_line = axis.plot(
            x_axis[fit_indices],
            group_amplitude * np.exp(-group_decay * x_axis[fit_indices]),
            color=fit_color,
            linewidth=2.4,
            label="Exponential fit",
        )[0]
        handles.append(fit_line)
    range_line = axis.axvline(
        x_axis[fit_indices[0]],
        color="black",
        linestyle="--",
        linewidth=1,
        label="Fit range",
    )
    axis.axvline(
        x_axis[fit_indices[-1]],
        color="black",
        linestyle="--",
        linewidth=1,
    )
    handles.append(range_line)
    _style_axis(
        axis,
        x_axis,
        xlabel,
        f"Exponential decay fit - Component #{component_index + 1}",
    )
    axis.legend(handles=handles, loc="upper right")
    figures.append(figure)

    if plot_all:
        for subject in range(nsubjects):
            figure, axis = plt.subplots(figsize=(8, 5), constrained_layout=True)
            spectrum_line = axis.plot(
                x_axis,
                eigenspectrum[:, subject],
                color=spectrum_color,
                linewidth=1.5,
                label="Eigenspectrum",
            )[0]
            handles = [spectrum_line]
            if np.isfinite(decay_coefficients[subject, 0]) and np.isfinite(
                amplitudes[subject, 0]
            ):
                fit_line = axis.plot(
                    x_axis[fit_indices],
                    amplitudes[subject, 0]
                    * np.exp(
                        -decay_coefficients[subject, 0] * x_axis[fit_indices]
                    ),
                    color=fit_color,
                    linewidth=2.4,
                    label="Exponential fit",
                )[0]
                handles.append(fit_line)
            range_line = axis.axvline(
                x_axis[fit_indices[0]],
                color="black",
                linestyle="--",
                linewidth=1,
                label="Fit range",
            )
            axis.axvline(
                x_axis[fit_indices[-1]],
                color="black",
                linestyle="--",
                linewidth=1,
            )
            handles.append(range_line)
            _style_axis(
                axis,
                x_axis,
                xlabel,
                f"Participant #{subject + 1} - Exponential decay fit "
                f"(component {component_index + 1})",
            )
            axis.legend(handles=handles, loc="upper right")
            figures.append(figure)
    return figures


def FREQNESS_ExponentialDK(
    FREQ: Any,
    *,
    which_comp: int | None = None,
    range2fit: ArrayLike | None = None,
    plot_all: bool = False,
    show: bool = True,
) -> tuple[FloatArray, FREQNESSExponentialGoodFit]:
    """Fit exponential eigenvalue decay across frequencies.

    The model is ``y = A * exp(-lambda * x)``. ``which_comp`` deliberately
    uses MATLAB-compatible one-based component numbering. Fits require at
    least three positive finite eigenvalues. R² is evaluated in the original,
    non-logarithmic eigenvalue space, matching the MATLAB implementation.

    Returns
    -------
    decayCoeff, goodFit
        ``decayCoeff`` and ``goodFit.R2`` both have shape
        ``(participants, 1)``.
    """
    for value, name in ((plot_all, "plot_all"), (show, "show")):
        if not isinstance(value, (bool, np.bool_)):
            raise TypeError(f"{name} must be a boolean")

    values = _eigenspectrum_array(FREQ)
    ncomponents, nfrequencies, nsubjects = values.shape
    component_index = _component_index(which_comp, ncomponents)
    x_axis, xlabel, has_frequencies = _frequency_axis(FREQ, nfrequencies)
    fit_indices = _fit_indices(x_axis, has_frequencies, range2fit)
    component_spectrum = values[component_index, :, :]

    decay_coefficients = np.full((nsubjects, 1), np.nan, dtype=float)
    amplitudes = np.full((nsubjects, 1), np.nan, dtype=float)
    r_squared = np.full((nsubjects, 1), np.nan, dtype=float)
    for subject in range(nsubjects):
        decay, amplitude, subject_r2 = _fit_exponential(
            x_axis[fit_indices],
            component_spectrum[fit_indices, subject],
        )
        decay_coefficients[subject, 0] = decay
        amplitudes[subject, 0] = amplitude
        r_squared[subject, 0] = subject_r2

    group_spectrum = _nanmean(component_spectrum, axis=1)
    group_decay, group_amplitude, _ = _fit_exponential(
        x_axis[fit_indices], group_spectrum[fit_indices]
    )
    _plot_decay(
        component_spectrum,
        x_axis,
        xlabel,
        fit_indices,
        component_index,
        decay_coefficients,
        amplitudes,
        group_decay,
        group_amplitude,
        plot_all=bool(plot_all),
    )
    if show:
        import matplotlib.pyplot as plt

        plt.show()

    return decay_coefficients, FREQNESSExponentialGoodFit(R2=r_squared)


__all__ = ["FREQNESS_ExponentialDK", "FREQNESSExponentialGoodFit"]

File: FREQNESS_Toolbox/FREQNESS_Functions/FREQNESS_ExponentialDK.m
function [decayCoeff, goodFit] = FREQNESS_ExponentialDK(FREQ, varargin)

% ========================================================================
%
%  FREQUENCY-RESOLVED NETWORK ESTIMATION TOOLBOX
%
%  If you use this toolbox, please cite:
%  Rosso, M., Fernández‐Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
%  Kringelbach, M. L., & Bonetti, L. (2025).
%  FREQ‐NESS Reveals the Dynamic Reconfiguration of Frequency‐Resolved Brain
%  Networks During Auditory Stimulation.
%  Advanced Science, 2413195.
%  https://doi.org/10.1002/advs.202413195
%
% ========================================================================
%  This function estimates how rapidly the eigenvalues associated with a
%  given component decay as a function of frequency. It does so by fitting
%  an exponentially decaying function to the eigenspectrum across
%  frequencies, either for all frequencies or within a user-defined
%  frequency range. 
%  
%  The decay rate can be interpreted as a compact index of
%  how strongly the component is dominated by low- vs. high-frequency
%  contributions, either across the entire spectrum or within a defined 
%  frequency range. 
% 
%  The coefficients in the output and can be compared across experimental 
%  conditions or groups of participants.
% ========================================================================

% ------------------------------------------------------------------------
%  INPUT ARGUMENTS:
% ------------------------------------------------------------------------
%
%  - FREQ : structure with fields
%           • FREQ.evals -> eigenvalue array [nComp x nFrex x (nSubs)]
%           • FREQ.frex  -> frequency vector [nFrex x 1] (optional but
%                            used for the x-axis in the visualization)
%
% - Optional arguments:
%
%   - which_comp : component index to analyse (default: 1)
% 
%   - range2fit  : frequency range to fit, expressed in Hz, e.g. [8 12].
%                  When empty, the exponential fit is performed across all
%                  available frequencies.
%
%   - plot_all   : logical flag (true/false). When true, the function
%                  produces additional figures for each individual
%                  participant, plotting their eigenspectrum and fitted
%                  exponential decay. By default, plot_all = false and
%                  only the grand-average fit is visualized.
%
% ------------------------------------------------------------------------
%  OUTPUT ARGUMENTS:
% ------------------------------------------------------------------------
%
%  - decayCoeff : Vector [nSubs x 1] containing the exponential decay
%                 coefficient (lambda) estimated for the selected
%                 component in each participant. Larger positive values
%                 correspond to a steeper decay of eigenvalues across
%                 frequencies (i.e., stronger dominance of low frequencies).
%
%  - goodFit    : Structure containing goodness-of-fit information for the
%                 exponential model. Currently:
%                   • goodFit.R2  -> [nSubs x 1] vector with the coefficient
%                                    of determination (R²) for each subject.
%                 These values quantify how well the exponential model
%                 captures the eigenspectrum shape within the fitted
%                 frequency range.
%
% ------------------------------------------------------------------------
%  REFERENCES:
% ------------------------------------------------------------------------
%
%  - Kudryavtsev et al., in preparation
%
%  - Rosso, M., Fernández‐Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
%    Kringelbach, M. L., & Bonetti, L. (2025).
%    FREQ‐NESS Reveals the Dynamic Reconfiguration of Frequency‐Resolved Brain
%    Networks During Auditory Stimulation. Advanced Science, 2413195.
%
% ------------------------------------------------------------------------
%  AUTHORS:
%  Mattia Rosso & Leonardo Bonetti
%  mattia.rosso@clin.au.dk
%  leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
%  Center for Music in the Brain, Aarhus University
%  Centre for Eudaimonia and Human Flourishing, Linacre College, University of Oxford
%  Aarhus (DK), Oxford (UK), 30/09/2025
%
% ========================================================================

%% Handle optional arguments (name-value pairs)

opts = struct('which_comp', [], 'range2fit', [], 'plot_all',false);
opts = parse_name_value_pairs(opts, varargin{:});

which_comp = opts.which_comp;
range2fit  = opts.range2fit;
plot_all   = opts.plot_all;   

%% Map inputs from FREQ 

if ~isstruct(FREQ)
    error('Input must be a struct with fields FREQ.evals (and optionally FREQ.frex).');
end
if ~isfield(FREQ,'evals') || isempty(FREQ.evals)
    error('FREQ.evals is missing or empty.');
end
eigenspectrum = FREQ.evals;

% Optional frequency axis for plotting
if isfield(FREQ,'frex') && ~isempty(FREQ.frex)
    frex = FREQ.frex(:);
end

%% Normalize input shape to [nComp x nFrex x nSubs]

if isvector(eigenspectrum)
    eigenspectrum = eigenspectrum(:);               % [nComp x 1]
end

% Get the number of dimensions in the input
nd = ndims(eigenspectrum);
if nd == 2
    % [nComp x nFrex] -> [nComp x nFrex x 1]
    eigenspectrum = reshape(eigenspectrum, size(eigenspectrum,1), size(eigenspectrum,2), 1);
elseif nd > 3
    error('eigenspectrum must be 1D, 2D, or 3D (nComp x nFrex x nSubs).');
end

% Define Ns
ncomps = size(eigenspectrum,1);
nfrex  = size(eigenspectrum,2);
nsubs  = size(eigenspectrum,3);

% Display for the user
fprintf('\nFREQNESS ExponentialDK: modelling eigenvalue decay from %.1f to %.1f Hz for %d participants.\n', ...
    range2fit(1), range2fit(end), nsubs);

%% Fit exponentially decaying function

% -------------------------------------------------------------------------
% 1) Select component
% -------------------------------------------------------------------------
if isempty(which_comp)
    disp('Component not specified. Defaulting to analyzing the 1st component.');
    which_comp = 1;
elseif ~isscalar(which_comp) || which_comp < 1 || which_comp > ncomps
    warning(['Input variable "which_comp" must be a scalar between 1 and ', num2str(ncomps), ...
             '. Defaulting to analyzing the 1st component.']);
    which_comp = 1;
end

% Extract eigenspectrum for the selected component: [nFrex x nSubs]
eig_comp = squeeze(eigenspectrum(which_comp,:,:)); 
if isvector(eig_comp)
    eig_comp = eig_comp(:); % [nFrex x 1] if single subject
end

% -------------------------------------------------------------------------
% 2) Define x-axis (frequency or index)
% -------------------------------------------------------------------------
if exist('frex','var') && numel(frex)==nfrex
    x_all = frex(:);
    xlab  = 'Frequency (Hz)';
else
    x_all = (1:nfrex).';
    xlab  = 'Frequency index';
end

% -------------------------------------------------------------------------
% 3) Map range2fit (in Hz) to indices to fit
% -------------------------------------------------------------------------
if isempty(range2fit)
    % Default: fit across all frequencies
    idx_range2fit = 1:nfrex;
else
    if ~exist('frex','var')
        error('range2fit was provided in Hz, but FREQ.frex is missing.');
    end
    if numel(range2fit) ~= 2
        error(['The variable range2fit must be a 2-element vector containing the boundaries ' ...
               'of the frequency range to fit the exponential decay, e.g. [8 12].']);
    end
    
    % Find closest 1st and last frequencies
    [~, idx_first] = min(abs(frex - range2fit(1)));
    [~, idx_last]  = min(abs(frex - range2fit(2)));
    idx_range2fit  = idx_first:idx_last;
    
    % Warn if requested boundaries are not exact matches
    tol = 1e-3; % Hz
    if any(~ismembertol(range2fit(:), frex(:), tol))
        warning('Some requested frequencies in range2fit are not present in FREQ.frex. Using closest matches instead.');
    end
end

% Frequencies (or indices) actually used for the fit
x_fit = x_all(idx_range2fit);

% -------------------------------------------------------------------------
% 4) Fit exponential decay per subject: y = A * exp(-lambda * x)
% -------------------------------------------------------------------------
decayCoeff = nan(nsubs,1);
goodFit.R2 = nan(nsubs,1);
A_sub      = nan(nsubs,1);   % store amplitudes for optional individual plots

for subi = 1:nsubs
    
    y = eig_comp(idx_range2fit, subi);
    
    % Keep only positive, finite values
    mask = y > 0 & isfinite(y);
    if sum(mask) < 3
        % Not enough data points to fit reliably
        continue
    end
    
    x_sub = x_fit(mask);
    y_sub = y(mask);
    
    % Linearize: log(y) = log(A) - lambda * x
    logy = log(y_sub);
    X    = [x_sub(:) ones(numel(x_sub),1)];
    
    beta = X \ logy;          % [slope; intercept]
    lambda = -beta(1);        % decay coefficient
    A      = exp(beta(2));    % amplitude (not returned, but used for plotting)
    
    % Store decay coefficient
    decayCoeff(subi) = lambda;
    A_sub(subi)      = A;
    
    % Compute R² in the original (non-log) space
    y_hat = A * exp(-lambda * x_sub);
    sse   = sum((y_sub - y_hat).^2);
    sst   = sum((y_sub - mean(y_sub)).^2);
    if sst > 0
        goodFit.R2(subi) = 1 - sse/sst;
    else
        goodFit.R2(subi) = NaN;
    end
end

% For visualization, also compute a group-level fit on the average eigenspectrum
mean_eig = mean(eig_comp, 2, 'omitnan');
y_group  = mean_eig(idx_range2fit);
mask_g   = y_group > 0 & isfinite(y_group);
if sum(mask_g) >= 3
    xg    = x_fit(mask_g);
    yg    = y_group(mask_g);
    logyg = log(yg);
    Xg    = [xg(:) ones(numel(xg),1)];
    betag = Xg \ logyg;
    lambda_group = -betag(1);
    A_group      = exp(betag(2));
else
    lambda_group = NaN;
    A_group      = NaN;
end

%% Visualize fit

% X-axis: use 'frex' if available, otherwise frequency index
if exist('frex','var') && numel(frex)==nfrex
    x = frex(:);
    xlab  = 'Frequency (Hz)';
else
    x = (1:nfrex).';
    xlab  = 'Frequency index';
end

% Mean ± SEM across subjects
if nsubs > 1
    avg_eig = mean(eig_comp, 2, 'omitnan');
    sem_eig = std(eig_comp, 0, 2, 'omitnan') ./ sqrt(nsubs);
else
    avg_eig = eig_comp;
    sem_eig = [];
end

% Parula colors
cmap    = .8*parula(256);
col_eig = cmap(1,:);      % eigenspectrum (purple)
col_fit = cmap(200,:);    % fitted curve (yellow)

figure('Color','w','Units','normalized','Position',[0.3 0.3 0.35 0.5]);
hold on

% SEM band
if nsubs > 1 && ~all(isnan(sem_eig))
    h_sem = fill([x; flipud(x)], [avg_eig - sem_eig; flipud(avg_eig + sem_eig)], ...
        [0 0 0], 'FaceAlpha', 0.1, 'EdgeColor', 'none');
end

% Mean eigenspectrum
h_eig = plot(x, avg_eig, 'Color', col_eig, 'LineWidth', 2);

% Fitted exponential decay
if ~isnan(A_group) && ~isnan(lambda_group)
    x_fit_line = x(idx_range2fit);
    y_fit_line = A_group * exp(-lambda_group * x_fit_line);
    h_fit = plot(x_fit_line, y_fit_line, 'Color', col_fit, 'LineWidth', 2.4);
end

% Range markers
x_min_fit = x(idx_range2fit(1));
x_max_fit = x(idx_range2fit(end));
h_min = xline(x_min_fit, 'k--', 'LineWidth', 1);
h_max = xline(x_max_fit, 'k--', 'LineWidth', 1);

xlabel(xlab, 'FontSize', 13, 'FontWeight', 'bold');
ylabel('Eigenvalue','FontSize',13,'FontWeight','bold');
title(sprintf('Exponential decay fit - Component #%d', which_comp), ...
      'FontSize',14,'FontWeight','bold');
set(gca,'FontSize',12,'LineWidth',1.2,'Box','off');
grid on; grid minor; xlim([min(x) max(x)]);
legend( ...
    [h_eig, h_fit, h_min], ...   % include only one dashed line (they are identical)
    {'Mean eigenspectrum', 'Exponential fit', 'Fit range'}, ...
    'Location','northeast', ...
    'FontSize',12);

% -------------------------------------------------------------------------
% Optional individual-subject plots
% -------------------------------------------------------------------------
if plot_all
    for subi = 1:nsubs
        
        y_full = eig_comp(:,subi);
        
        figure('Color','w','Units','normalized','Position',[0.3 0.3 0.35 0.5]);
        hold on
        
        % Subject eigenspectrum
        h_eig_s = plot(x, y_full, 'Color', col_eig, 'LineWidth', 1.5);
      
        % Fit
        if ~isnan(A_sub(subi)) && ~isnan(decayCoeff(subi))
            x_fit_line = x(idx_range2fit);
            y_fit_line = A_sub(subi) * exp(-decayCoeff(subi) * x_fit_line);
            h_fit_s = plot(x_fit_line, y_fit_line, 'Color', col_fit, 'LineWidth', 2.4);
        end
        
        % Range
        h_min_s = xline(x_min_fit, 'k--', 'LineWidth', 1);
        xline(x_max_fit, 'k--', 'LineWidth', 1);
        
        xlabel(xlab,'FontSize',13,'FontWeight','bold');
        ylabel('Eigenvalue','FontSize',13,'FontWeight','bold');
        title(sprintf('Subject %d – exponential decay fit (component %d)', subi, which_comp), ...
              'FontSize',14,'FontWeight','bold');
        set(gca,'FontSize',12,'LineWidth',1.2,'Box','off');
        grid on; grid minor; xlim([min(x) max(x)]);

        % Legend (per subject)
        if exist('h_fit_s','var')
            legend([h_eig_s, h_fit_s, h_min_s], ...
                {'Eigenspectrum','Exponential fit','Fit range'}, ...
                'Location','northeast','FontSize',12);
        else
            legend([h_eig_s, h_min_s], ...
                {'Eigenspectrum','Fit range'}, ...
                'Location','northeast','FontSize',12);
        end

    end
end

end


%% Helper Function: Parse Name-Value Pairs
function opts = parse_name_value_pairs(opts, varargin)
% Simple name-value parser used to handle optional arguments.
% Example:
%   opts = struct('which_comp', [], 'range2fit', [], 'plot_all', false);
%   opts = parse_name_value_pairs(opts, 'which_comp', 2, 'range2fit', [8 12], 'plot_all', true);

if mod(length(varargin), 2) ~= 0
    error('Arguments must be given as name-value pairs.');
end

for i = 1:2:length(varargin)
    name = lower(varargin{i});
    if isfield(opts, name)
        opts.(name) = varargin{i+1};
    else
        error(['Unrecognized argument: ', name]);
    end
end
end
